### PR TITLE
Multithreading support and examples

### PR DESCRIPTION
**Pull request by filipinascimento**
These examples illustrate how to interact with fury using two different threads.

Offscreen example uses the main thread to render the scene, while a second thread keeps saving a snapshot at random intervals.

The other example opens a window in the main thread and updates the scene form a second thread.
Note that the `iren` loop had to be done manually (`showm.start()` will not work in this case). 

This is a work in progress PR.

### CODE REVIEW

**Comment by Garyfallidis** [comment]
@filipinascimento the CI fails are not relevant to your code. Can you remove all unrelated or old examples from this PR?

File: fury/window.py [hunk 1]
```diff
@@ -1190,3 +1264,21 @@ def gl_set_subtractive_blending(gl_state):
...
+    Parameters
+    ----------
+    window : vtkRenderWindow
+
+    """
+    # Once release current context is available:
+    # https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8418
+    try:
+        window.ReleaseCurrent()
+    except AttributeError:
```
Comment (Garyfallidis) [inline fury/window.py:1281]: Perhaps remove this part?

File: fury/window.py [hunk 2]
```diff
@@ -371,15 +375,48 @@ def render(self):
...
+        ----------
+        multithreaded : bool
+            Whether to use multithreading. (Default False)
+        desired_fps : int
+            Desired frames per second when using multithreading is enabled.
+            (Default 60)
+
+        """
+        try:
+            if(multithreaded):
```
Comment (skoudoro) [inline fury/window.py:398]: no parenthesis

File: fury/window.py [hunk 3]
```diff
@@ -371,15 +375,48 @@ def render(self):
...
+                while self.iren.GetDone() is False:
+                    startTime = time.perf_counter()
```
Comment (skoudoro) [inline fury/window.py:400]: variable name:  `start_time`

File: fury/window.py [hunk 4]
```diff
@@ -371,15 +375,48 @@ def render(self):
...
+                    self.lock()
+                    self.window.MakeCurrent()
+                    self.iren.ProcessEvents()  # Check if we can really do that
+                    self.window.Render()
+                    release_context(self.window)
+                    self.release_lock()
+                    endTime = time.perf_counter()
```
Comment (skoudoro) [inline fury/window.py:407]: variable name: `end_time`

File: fury/window.py [hunk 5]
```diff
@@ -371,15 +375,48 @@ def render(self):
...
+                    # throttle to 60fps to avoid busy wait
+                    timePerFrame = 1.0/desired_fps
```
Comment (skoudoro) [inline fury/window.py:409]: variable name: `time_per_frame`

File: fury/window.py [hunk 6]
```diff
@@ -371,15 +375,48 @@ def render(self):
...
+                    if endTime - startTime < timePerFrame:
+                        time.sleep(timePerFrame - (endTime - startTime))
             else:
-                self.window.SetWindowName(self.title)
-            self.iren.Start()
+                self.render()
+                if self.title.upper() == "FURY":
+                    self.window.SetWindowName(self.title + " " + fury_version)
+                else:
+                    self.window.SetWindowName(self.title)
```
Comment (skoudoro) [inline fury/window.py:417]: you need to keep this. otherwise, we can not setup the window name

File: fury/window.py [hunk 7]
```diff
@@ -400,6 +437,43 @@ def start(self):
...
+
+    def lock_current(self):
+        """Lock the render window and acuiqre the current context and
+        check if the lock was sucessfully acquired.
+
+        Returns
+        -------
+        sucessful : bool
+            Returns if the lock was acquired."""
+        if(self.is_done()):
```
Comment (skoudoro) [inline fury/window.py:456]: remove parenthesis

File: fury/window.py [hunk 8]
```diff
@@ -400,6 +437,43 @@ def start(self):
...
+                return False
+
+    def release_current(self):
+        """Release the window context and lock of the render window."""
+        release_context(self.window)
+        self.release_lock()
+
+    def wait(self):
+        """ Wait for thread to finish. """
+        if(self.thread):
```
Comment (skoudoro) [inline fury/window.py:474]: remove parenthesis

**Review by skoudoro — COMMENTED** [review]
Hi @filipinascimento,

Thank you for the update! See below for some comments

**Comment by Garyfallidis** [comment]
Ping @filipinascimento ! It would be great to have this merged asap. Please make the required changes.

**Comment by Garyfallidis** [comment]
Ping @filipinascimento !!! Let's get this PR to move forward! TY!

**Review by Garyfallidis — CHANGES_REQUESTED** [review]
@filipinascimento It seems all tests are passing now. Can you also simply fix the PEP8 and doc issues?

File: fury/tests/test_window.py [hunk 9]
```diff
@@ -11,6 +13,11 @@
 from fury.decorators import skip_osx, skip_win
 from fury import shaders
 
+from fury.lib import (Renderer, Volume, Actor2D, InteractorEventRecorder,
+                      InteractorStyleImage, InteractorStyleTrackballCamera,
+                      RenderWindow, RenderWindowInteractor, RenderLargeImage,
+                      WindowToImageFilter, Command)
```
Comment (skoudoro) [inline fury/tests/test_window.py:19]: Not sure what is the goal of all this additional import since they are not used

File: fury/window.py [hunk 10]
```diff
@@ -1,5 +1,8 @@
 # -*- coding: utf-8 -*-
 
+from sys import platform
```
Comment (skoudoro) [inline fury/window.py:3]: pep8: not used

**Review by skoudoro — COMMENTED** [review]
Hi @filipinascimento,

See below for some comments. The tests are failing on macOS but it works on my local machine. We need to figure out what is the issue.
the error is `illegal instruction 4` which is usually due to an architecture issue. The operating system in our CIs is macOS 11.6.5 with:
- Included Software: https://github.com/actions/virtual-environments/blob/macOS-11/20220508.1/images/macos/macos-11-Readme.md
- Image Release: https://github.com/actions/virtual-environments/releases/tag/macOS-11%2F20220508.1
maybe you can try via a virtual machine. 

Thank you in advance for your feedback

File: fury/tests/test_window.py [hunk 11]
```diff
@@ -1,4 +1,6 @@
-from fury.utils import remove_observer_from_actor
+import time
+from threading import Thread
+from fury.utils import remove_observer_from_actor, rotate, update_actor, vertices_from_actor,shallow_copy
```
[the remark this answers is not in the document]
Comment (filipinascimento) [reply fury/tests/test_window.py:3]: Merge problem, fixed.

File: fury/tests/test_window.py (hunk 9 above)
Comment (filipinascimento) [reply fury/tests/test_window.py:19]: Probably a problem with previous merge. Fixed now

File: fury/window.py (hunk 10 above)
Comment (filipinascimento) [reply fury/window.py:3]: fixed

File: fury/window.py [hunk 12]
```diff
@@ -434,6 +472,42 @@ def start(self):
...
+    def lock(self):
+        """Lock the render window."""
+        self.mutex.acquire()
+
+    def release_lock(self):
+        """Release the lock of the render window."""
+        self.mutex.release()
+
+    def lock_current(self):
+        """Lock the render window and acuiqre the current context and
```
[the remark this answers is not in the document]
Comment (filipinascimento) [reply fury/window.py:484]: fixed

File: fury/window.py [hunk 13]
```diff
@@ -434,6 +472,42 @@ def start(self):
...
+        check if the lock was sucessfully acquired.
+
+        Returns
+        -------
+        sucessful : bool
+            Returns if the lock was acquired."""
+        if self.is_done():
+            return False
+        try:
+            self.window
```
[the remark this answers is not in the document]
Comment (filipinascimento) [reply fury/window.py:494]: Fixed. now using hasattr instead. This checks if showm is still valid.

File: fury/tests/test_thread.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+def test_multithreading():
+    xyz = 10 * (np.random.random((100, 3))-0.5)
+    colors = np.random.random((100, 4))
+    radii = np.random.random(100) + 0.5
+
+    scene = window.Scene()
+    sphere_actor = actor.sphere(centers=xyz,
+                                colors=colors,
+                                radii=radii,
+                                use_primitive = False)
```
[the remark this answers is not in the document]
Comment (filipinascimento) [reply fury/tests/test_thread.py:20]: Fixed

File: fury/tests/test_thread.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    scene.add(sphere_actor)
+
+    # Preparing the show manager as usual
+    showm = window.ShowManager(scene,
+                            size=(900, 768), reset_camera=False,
+                            order_transparent=True)
```
[the remark this answers is not in the document]
Comment (filipinascimento) [reply fury/tests/test_thread.py:26]: Fixed

**Comment by filipinascimento** [comment]
> Hi @filipinascimento,
> 
> See below for some comments. The tests are failing on macOS but it works on my local machine. We need to figure out what is the issue. the error is `illegal instruction 4` which is usually due to an architecture issue. The operating system in our CIs is macOS 11.6.5 with:
> 
> * Included Software: https://github.com/actions/virtual-environments/blob/macOS-11/20220508.1/images/macos/macos-11-Readme.md
> * Image Release: https://github.com/actions/virtual-environments/releases/tag/macOS-11%2F20220508.1
>   maybe you can try via a virtual machine.
> 
> Thank you in advance for your feedback

Not sure how to address this problem. It happens in the threading.py when calling thread.run(), maybe related to https://github.com/tqdm/tqdm/issues/1037 . I'm not sure why threading.py would have any illegal instructions, unless that macOS VM runs in unsupported CPU microarchitecture. It is not clear which CPU model they are using in the macOS VMs.

**Comment by m-agour** [comment]
Hello @filipinascimento,
About the FPS issue not being accurate while using threads, it's as we suspected an OS related problem. 
Linux kernels have minimal sleep interval down to 1ms. But it's higher in windows (14ms).
This got fixed in Python 3.11 and the  `time.sleep()` has the same accuracy regarding the OS, but since we need to maintain older versions and I think VTK is yet to be ported to python 3.11, we really can't make use of that.
A way to fix this for Windows so is setting the minimum timer resolution to 1ms:
```
import ctypes
import sys

if sys.platform == 'win32' and sys.version_info[1] < 11:
    winmm = ctypes.WinDLL('winmm')
    winmm.timeBeginPeriod(1)
```
This fixed the problem for me.

**Comment by filipinascimento** [comment]
> Hello @filipinascimento, About the FPS issue not being accurate while using threads, it's as we suspected an OS related problem. Linux kernels have minimal sleep interval down to 1ms. But it's higher in windows (14ms). This got fixed in Python 3.11 and the `time.sleep()` has the same accuracy regarding the OS, but since we need to maintain older versions and I think VTK is yet to be ported to python 3.11, we really can't make use of that. A way to fix this for Windows so is setting the minimum timer resolution to 1ms:
> 
> ```
> import ctypes
> import sys
> 
> if sys.platform == 'win32' and sys.version_info[1] < 11:
>     winmm = ctypes.WinDLL('winmm')
>     winmm.timeBeginPeriod(1)
> ```
> 
> This fixed the problem for me.

Thanks,  this may work as a temporary solution.

**Pull request merged**